File: combinatorics/resource/LCHP/problem_LTDT_code/bai_11/bai11.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <set>
#include <iomanip>
#include <algorithm>

using namespace std;

class SimpleDFS {
private:
    int numVertices;
    vector<vector<int>> adjList;
    vector<bool> visited;
    vector<int> startTime;
    vector<int> finishTime;
    vector<int> parent;
    vector<int> dfsOrder;
    int timeCounter;
    
    // Phan loai canh
    enum EdgeType { TREE, BACK, FORWARD, CROSS };
    vector<tuple<int, int, EdgeType>> edges;
    
public:
    SimpleDFS(int n) : numVertices(n), timeCounter(0) {
        adjList.resize(n);
        resetArrays();
    }
    
    void resetArrays() {
        visited.assign(numVertices, false);
        startTime.assign(numVertices, 0);
        finishTime.assign(numVertices, 0);
        parent.assign(numVertices, -1);
        dfsOrder.clear();
        edges.clear();
        timeCounter = 0;
    }
    
    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u); // Do thi vo huong
    }
    
    // DFS de quy
    void dfsRecursive(int u) {
        visited[u] = true;
        startTime[u] = ++timeCounter;
        dfsOrder.push_back(u);
        
        cout << "Tham dinh " << u << " (start: " << startTime[u] << ")" << endl;
        
        // Duyet cac dinh ke
        for (int v : adjList[u]) {
            if (!visited[v]) {
                parent[v] = u;
                edges.push_back({u, v, TREE});
                cout << "  Canh cay: " << u << " -> " << v << endl;
                dfsRecursive(v);
            } else {
                // Phan loai canh
                if (startTime[v] < startTime[u] && finishTime[v] == 0) {
                    edges.push_back({u, v, BACK});
                    cout << "  Canh nguoc: " << u << " -> " << v << " (chu trinh!)" << endl;
                } else if (startTime[v] > startTime[u]) {
                    edges.push_back({u, v, FORWARD});
                    cout << "  Canh xuoi: " << u << " -> " << v << endl;
                } else {
                    edges.push_back({u, v, CROSS});
                    cout << "  Canh cheo: " << u << " -> " << v << endl;
                }
            }
        }
        
        finishTime[u] = ++timeCounter;
        cout << "Ket thuc dinh " << u << " (finish: " << finishTime[u] << ")" << endl;
    }
    
    // DFS khong de quy
    void dfsIterative(int start) {
        resetArrays();
        
        stack<int> stk;
        stack<int> path; // De theo doi duong di
        
        cout << "DFS khong de quy tu dinh " << start << ":" << endl;
        
        stk.push(start);
        
        while (!stk.empty()) {
            int u = stk.top();
            stk.pop();
            
            if (!visited[u]) {
                visited[u] = true;
                startTime[u] = ++timeCounter;
                dfsOrder.push_back(u);
                
                cout << "Tham dinh " << u << " (start: " << startTime[u] << ")" << endl;
                
                // Them cac dinh ke vao stack (thu tu nguoc lai)
                vector<int> neighbors = adjList[u];
                sort(neighbors.rbegin(), neighbors.rend());
                
                for (int v : neighbors) {
                    if (!visited[v]) {
                        stk.push(v);
                        if (parent[v] == -1) {
                            parent[v] = u;
                        }
                    }
                }
            }
        }
        cout << endl;
    }
    
    // DFS toan bo do thi
    void dfsComplete() {
        resetArrays();
        
        cout << "DFS toan bo do thi:" << endl;
        int componentCount = 0;
        
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                componentCount++;
                cout << "\nThanh phan lien thong " << componentCount << ":" << endl;
                dfsRecursive(i);
            }
        }
        
        cout << "\nTong so thanh phan lien thong: " << componentCount << endl;
        analyzeCycles();
    }
    
    // Phan tich chu trinh
    void analyzeCycles() {
        bool hasCycle = false;
        
        cout << "\nPhan tich canh:" << endl;
        for (auto& edge : edges) {
            int u = get<0>(edge);
            int v = get<1>(edge);
            EdgeType type = get<2>(edge);
            
            string typeStr;
            switch (type) {
                case TREE: typeStr = "Cay"; break;
                case BACK: typeStr = "Nguoc"; hasCycle = true; break;
                case FORWARD: typeStr = "Xuoi"; break;
                case CROSS: typeStr = "Cheo"; break;
            }
            
            cout << "Canh (" << u << "," << v << "): " << typeStr << endl;
        }
        
        cout << "\nDo thi co chu trinh: " << (hasCycle ? "Co" : "Khong") << endl;
    }
    
    // Tim tat ca duong di tu start den end
    void findAllPaths(int start, int end) {
        vector<vector<int>> allPaths;
        vector<int> currentPath;
        vector<bool> tempVisited(numVertices, false);
        
        findAllPathsHelper(start, end, currentPath, allPaths, tempVisited);
        
        cout << "\nTat ca duong di tu " << start << " den " << end << ":" << endl;
        for (int i = 0; i < allPaths.size(); i++) {
            cout << "Duong " << i + 1 << ": ";
            for (int j = 0; j < allPaths[i].size(); j++) {
                cout << allPaths[i][j];
                if (j < allPaths[i].size() - 1) cout << " -> ";
            }
            cout << endl;
        }
        
        if (allPaths.empty()) {
            cout << "Khong co duong di tu " << start << " den " << end << endl;
        }
    }
    
private:
    void findAllPathsHelper(int current, int end, vector<int>& path, 
                           vector<vector<int>>& allPaths, vector<bool>& tempVisited) {
        path.push_back(current);
        tempVisited[current] = true;
        
        if (current == end) {
            allPaths.push_back(path);
        } else {
            for (int neighbor : adjList[current]) {
                if (!tempVisited[neighbor]) {
                    findAllPathsHelper(neighbor, end, path, allPaths, tempVisited);
                }
            }
        }
        
        path.pop_back();
        tempVisited[current] = false;
    }
    
public:
    // Sap xep topo (chi cho do thi co huong)
    vector<int> topologicalSort() {
        resetArrays();
        stack<int> topoStack;
        
        // Goi DFS cho tat ca dinh
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                topoDFS(i, topoStack);
            }
        }
        
        vector<int> result;
        while (!topoStack.empty()) {
            result.push_back(topoStack.top());
            topoStack.pop();
        }
        
        return result;
    }
    
private:
    void topoDFS(int u, stack<int>& topoStack) {
        visited[u] = true;
        
        for (int v : adjList[u]) {
            if (!visited[v]) {
                topoDFS(v, topoStack);
            }
        }
        
        topoStack.push(u);
    }
    
public:
    // In thong tin thoi gian
    void printTimes() {
        cout << "\nThong tin thoi gian DFS:" << endl;
        cout << setw(6) << "Dinh" << setw(12) << "Start" << setw(12) << "Finish" 
             << setw(10) << "Cha" << endl;
        cout << string(40, '-') << endl;
        
        for (int i = 0; i < numVertices; i++) {
            cout << setw(6) << i 
                 << setw(12) << startTime[i]
                 << setw(12) << finishTime[i]
                 << setw(10) << parent[i] << endl;
        }
        cout << endl;
    }
    
    void printDFSTree() {
        cout << "Cay DFS:" << endl;
        for (int i = 0; i < numVertices; i++) {
            if (parent[i] != -1) {
                cout << parent[i] << " -> " << i << endl;
            }
        }
        cout << endl;
    }
    
    void printGraph() {
        cout << "Bieu dien do thi don:" << endl;
        for (int i = 0; i < numVertices; i++) {
            cout << i << ": ";
            for (int neighbor : adjList[i]) {
                cout << neighbor << " ";
            }
            cout << endl;
        }
        cout << endl;
    }
    
    // Kiem tra tinh lien thong
    bool isConnected() {
        dfsRecursive(0);
        
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                return false;
            }
        }
        return true;
    }
};

// Demo cho do thi don
void demonstrateSimpleDFS() {
    cout << "=== DEMO DFS CHO DO THI DON ===" << endl << endl;
    
    SimpleDFS graph(6);
    
    // Tao do thi mau
    graph.addEdge(0, 1);
    graph.addEdge(0, 2);
    graph.addEdge(1, 3);
    graph.addEdge(2, 4);
    graph.addEdge(3, 5);
    graph.addEdge(4, 5);
    graph.addEdge(1, 2); // Tao chu trinh
    
    graph.printGraph();
    
    // DFS de quy
    cout << "1. DFS DE QUY:" << endl;
    graph.dfsComplete();
    graph.printTimes();
    graph.printDFSTree();
    
    // DFS khong de quy
    cout << "2. DFS KHONG DE QUY:" << endl;
    graph.dfsIterative(0);
    
    // Tim tat ca duong di
    graph.findAllPaths(0, 5);
    
    // Kiem tra lien thong
    cout << "\nDo thi co lien thong: " 
         << (graph.isConnected() ? "Co" : "Khong") << endl;
}

int main() {
    demonstrateSimpleDFS();
    return 0;
}

File: combinatorics/resource/LCHP/problem_LTDT_code/bai_11/bai11.py
from collections import defaultdict
from typing import List, Tuple, Dict, Set
import sys

class SimpleDFS:
    """DFS cho đồ thị đơn"""
    
    def __init__(self, num_vertices: int):
        self.num_vertices = num_vertices
        self.adj_list = defaultdict(list)
        self.reset_arrays()
    
    def reset_arrays(self):
        self.visited = [False] * self.num_vertices
        self.start_time = [0] * self.num_vertices
        self.finish_time = [0] * self.num_vertices
        self.parent = [-1] * self.num_vertices
        self.dfs_order = []
        self.time_counter = 0
        self.edges = []  # (u, v, type)
    
    def add_edge(self, u: int, v: int):
        """Thêm cạnh vào đồ thị vô hướng"""
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)
    
    def dfs_recursive(self, u: int):
        """DFS đệ quy"""
        self.visited[u] = True
        self.time_counter += 1
        self.start_time[u] = self.time_counter
        self.dfs_order.append(u)
        
        print(f"Thăm đỉnh {u} (start: {self.start_time[u]})")
        
        # Duyệt các đỉnh kề
        for v in sorted(self.adj_list[u]):
            if not self.visited[v]:
                self.parent[v] = u
                self.edges.append((u, v, "TREE"))
                print(f"  Cạnh cây: {u} -> {v}")
                self.dfs_recursive(v)
            else:
                # Phân loại cạnh
                if (self.start_time[v] < self.start_time[u] and 
                    self.finish_time[v] == 0):
                    self.edges.append((u, v, "BACK"))
                    print(f"  Cạnh ngược: {u} -> {v} (chu trình!)")
                elif self.start_time[v] > self.start_time[u]:
                    self.edges.append((u, v, "FORWARD"))
                    print(f"  Cạnh xuôi: {u} -> {v}")
                else:
                    self.edges.append((u, v, "CROSS"))
                    print(f"  Cạnh chéo: {u} -> {v}")
        
        self.time_counter += 1
        self.finish_time[u] = self.time_counter
        print(f"Kết thúc đỉnh {u} (finish: {self.finish_time[u]})")
    
    def dfs_iterative(self, start: int):
        """DFS không đệ quy sử dụng stack"""
        self.reset_arrays()
        
        stack = [start]
        print(f"DFS không đệ quy từ đỉnh {start}:")
        
        while stack:
            u = stack.pop()
            
            if not self.visited[u]:
                self.visited[u] = True
                self.time_counter += 1
                self.start_time[u] = self.time_counter
                self.dfs_order.append(u)
                
                print(f"Thăm đỉnh {u} (start: {self.start_time[u]})")
                
                # Thêm các đỉnh kề vào stack (thứ tự ngược lại)
                neighbors = sorted(self.adj_list[u], reverse=True)
                for v in neighbors:
                    if not self.visited[v]:
                        stack.append(v)
                        if self.parent[v] == -1:
                            self.parent[v] = u
        print()
    
    def dfs_complete(self):
        """DFS toàn bộ đồ thị"""
        self.reset_arrays()
        
        print("DFS toàn bộ đồ thị:")
        component_count = 0
        
        for i in range(self.num_vertices):
            if not self.visited[i]:
                component_count += 1
                print(f"\nThành phần liên thông {component_count}:")
                self.dfs_recursive(i)
        
        print(f"\nTổng số thành phần liên thông: {component_count}")
        self.analyze_cycles()
    
    def analyze_cycles(self):
        """Phân tích chu trình"""
        has_cycle = False
        
        print("\nPhân tích cạnh:")
        for u, v, edge_type in self.edges:
            print(f"Cạnh ({u},{v}): {edge_type}")
            if edge_type == "BACK":
                has_cycle = True
        
        print(f"\nĐồ thị có chu trình: {'Có' if has_cycle else 'Không'}")
    
    def find_all_paths(self, start: int, end: int) -> List[List[int]]:
        """Tìm tất cả đường đi từ start đến end"""
        all_paths = []
        current_path = []
        temp_visited = [False] * self.num_vertices
        
        self._find_all_paths_helper(start, end, current_path, all_paths, temp_visited)
        
        print(f"\nTất cả đường đi từ {start} đến {end}:")
        for i, path in enumerate(all_paths, 1):
            path_str = ' -> '.join(map(str, path))
            print(f"Đường {i}: {path_str}")
        
        if not all_paths:
            print(f"Không có đường đi từ {start} đến {end}")
        
        return all_paths
    
    def _find_all_paths_helper(self, current: int, end: int, path: List[int], 
                              all_paths: List[List[int]], temp_visited: List[bool]):
        """Helper function for finding all paths"""
        path.append(current)
        temp_visited[current] = True
        
        if current == end:
            all_paths.append(path.copy())
        else:
            for neighbor in self.adj_list[current]:
                if not temp_visited[neighbor]:
                    self._find_all_paths_helper(neighbor, end, path, all_paths, temp_visited)
        
        path.pop()
        temp_visited[current] = False
    
    def topological_sort(self) -> List[int]:
        """Sắp xếp topo (chỉ cho đồ thị có hướng)"""
        self.reset_arrays()
        topo_stack = []
        
        # Gọi DFS cho tất cả đỉnh
        for i in range(self.num_vertices):
            if not self.visited[i]:
                self._topo_dfs(i, topo_stack)
        
        return topo_stack[::-1]
    
    def _topo_dfs(self, u: int, topo_stack: List[int]):
        """DFS helper for topological sort"""
        self.visited[u] = True
        
        for v in self.adj_list[u]:
            if not self.visited[v]:
                self._topo_dfs(v, topo_stack)
        
        topo_stack.append(u)
    
    def print_times(self):
        """In thông tin thời gian DFS"""
        print("\nThông tin thời gian DFS:")
        print(f"{'Đỉnh':>6} {'Start':>12} {'Finish':>12} {'Cha':>10}")
        print("-" * 40)
        
        for i in range(self.num_vertices):
            print(f"{i:>6} {self.start_time[i]:>12} {self.finish_time[i]:>12} {self.parent[i]:>10}")
        print()
    
    def print_dfs_tree(self):
        """In cây DFS"""
        print("Cây DFS:")
        for i in range(self.num_vertices):
            if self.parent[i] != -1:
                print(f"{self.parent[i]} -> {i}")
        print()
    
    def print_graph(self):
        """In biểu diễn đồ thị"""
        print("Biểu diễn đồ thị đơn:")
        for i in range(self.num_vertices):
            neighbors = ' '.join(map(str, sorted(self.adj_list[i])))
            print(f"{i}: {neighbors}")
        print()
    
    def is_connected(self) -> bool:
        """Kiểm tra tính liên thông"""
        self.reset_arrays()
        self.dfs_recursive(0)
        
        return all(self.visited)

# Demo cho đồ thị đơn
def demonstrate_simple_dfs():
    print("=== DEMO DFS CHO ĐỒ THỊ ĐƠN ===\n")
    
    graph = SimpleDFS(6)
    
    # Tạo đồ thị mẫu
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 3)
    graph.add_edge(2, 4)
    graph.add_edge(3, 5)
    graph.add_edge(4, 5)
    graph.add_edge(1, 2)  # Tạo chu trình
    
    graph.print_graph()
    
    # DFS đệ quy
    print("1. DFS ĐỆ QUY:")
    graph.dfs_complete()
    graph.print_times()
    graph.print_dfs_tree()
    
    # DFS không đệ quy
    print("2. DFS KHÔNG ĐỆ QUY:")
    graph.dfs_iterative(0)
    
    # Tìm tất cả đường đi
    graph.find_all_paths(0, 5)
    
    # Kiểm tra liên thông
    print(f"\nĐồ thị có liên thông: {'Có' if graph.is_connected() else 'Không'}")

if __name__ == "__main__":
    demonstrate_simple_dfs()